cpp/include/cudf/strings/string_view.cuh

```diff
@@ -30,6 +30,8 @@
 #include <thrust/execution_policy.h>
 #endif
 
+#include <algorithm>
```

Comment:
There is a call to `std::max` in this file. The include was missed is causing compile errors in a different repo/library.

Reply:
I could not figure out to call CUDA's min. I kept getting compile errors that `min` is not found.

Reply:
Strange. I don't immediately see why that would be the case, but I'm fine with leaving this as-is and including `<algorithm>` rather than spending more time on it.

Reply:
It seems that there is only `cuda::std::fmin` in `<cuda/std/cmath>` for floating points (https://github.com/NVIDIA/libcudacxx/blob/main/include/cuda/std/detail/libcxx/include/cmath#L203). Since `min` is simple enough can we just implement it again like `auto result = x < a ? x : a;`?

Reply:
Just for the record in case this comes up again and I do a search in the repo for it later, here is the error I would get:
```
/cudf/cpp/include/cudf/strings/string_view.cuh: In member function 'cudf::size_type cudf::string_view::find_impl(const char*, cudf::size_type, cudf::size_type, cudf::size_type) const':
/cudf/cpp/include/cudf/strings/string_view.cuh:343:33: error: there are no arguments to 'min' that depend on a template parameter, so a declaration of 'min' must be available [-fpermissive]
  343 |   auto const epos = byte_offset(min(pos + count, nchars));
      |                                 ^~~
/cudf/cpp/include/cudf/strings/string_view.cuh:343:33: note: (if you use '-fpermissive', G++ will accept your code, but allowing the use of an undeclared name is deprecated)
```
Seems like a declaration is missing perhaps.

Reply:
Sounds like gcc is trying to compile a `__device__` function. Is something including this `.cuh` header in host code?

Reply:
Yes, here is a bit more of the error stack
```
/cudf/cpp/src/binaryop/compiled/util.cpp
In file included from /cudf/cpp/include/cudf/column/column_device_view.cuh:22,
                 from /cudf/cpp/src/binaryop/compiled/util.cpp:20:
/cudf/cpp/include/cudf/strings/string_view.cuh: In member function 'cudf::size_type cudf::string_view::find_impl(const char*, cudf::size_type, cudf::size_type, cudf::size_type) const':
/cudf/cpp/include/cudf/strings/string_view.cuh:343:33: error: there are no arguments to 'min' that depend on a template parameter, so a declaration of 'min' must be available [-fpermissive]
  343 |   auto const epos = byte_offset(min(pos + count, nchars));
      |                                 ^~~
/cudf/cpp/include/cudf/strings/string_view.cuh:343:33: note: (if you use '-fpermissive', G++ will accept your code, but allowing the use of an undeclared name is deprecated)
```
The `src/binaryop/util.cpp` is including some `.cuh` headers so I guess the c++ compiler is choking on `min`.

Reply:
This feels like we're tugging another IWYU thread here... I'm assuming something else is depending on those cuh headers from util.cpp

Reply:
Glad we identified the cause. I vote to fix that in a separate PR.

Reply:
`rg "include.*cuh" --type cpp` shows we have tons of cpp files including CUDA headers. Tests including `<cudf/detail/iterator.cuh>` is a big culprit. I'd like to address this systematically if possible, because otherwise we are prone to mystery errors like the one David experienced.

Reply:
> Glad we identified the cause. I vote to fix that in a separate PR.

Agreed, we can aim to eliminate this systematically in a separate PR

> Sometimes we need to include .cuh files from host-only code for various reasons

I would be curious to know why. Offhand to me this would indicate a `cuh` file that should be split into two parts, with the host-usable parts moved to a separate `hpp` file.

Reply:
@davidwendt and I investigated further. This is a minimal reproduction of the error (which only requires a templated function with a missing definition): https://godbolt.org/z/fvKbhrKav

The `min` function is defined by CUDA. However, `cuda_runtime.h` only defines `min` in `crt/math_functions.h` if `__CUDACC__` is defined, so the host compiler won't see a definition for `min`.

However, the host compiler _does_ see the `__device__` function as just another function with `__attribute__((device))` which is ignored as an unrecognized attribute. See this for a demo of what the host compiler sees (using definitions from `cuda_runtime.h`, which includes `crt/host_defines.h`): https://godbolt.org/z/sGvWc8794

In summary, we are required to make all `__device__` functions compatible with the host compiler if the `.cuh` header containing them is included in host code (`.cpp`).

Reply:
Could you elaborate on why there are cases where we _need_ to include cuh files from host-only code? I'm not advocating making any further changes in this PR, just trying to understand the restriction here.

Reply:
I don't think it is _need_ as much as we _can_. There are classes like `column_device_view` that have a mixture of host and device member functions. It would be possible to create a `column_device_view` in a .cpp file (and use it's host member functions) and then maybe pass it to a function defined in a `.cu` file that calls the device member functions.

Also, I think the `iterators.cuh` header defines many iterators that can be used in either host or device code. We use this header in many .cpp test files. Perhaps the iterators need to be separated out.

Reply:
I've discussed this with @jrhemstad in the past and I believe we concluded the _need_ comes from the one definition rule (ODR). We have some classes with host and device code that must be declared in a `.cuh` but the host-side code needs to be used in `.cpp` files. However, there are definitely places where we can and should split `.cuh` headers into `.hpp` and `.cuh` components.

Reply:
Ah yes I didn't consider classes with both host and device methods. However, that seems odd to me. How does such a class even compile? The host compiler shouldn't know what `__device__` is, so if I `#include` a file containing `__device__` anywhere in a cpp file compiled with the host compiler, I would expect compiler errors unless those blocks are wrapped in preprocessor directives for conditional compilation. However, I see that for example `column_device_view_base::is_valid` is not wrapped in an `#ifdef __CUDACC__`, and yet clearly it is working for the utils.cpp->column_device_view.cuh file case. What am I missing?

Reply:
If `#include <cudf_runtime.h>` is somewhere in the .cpp TU then it does some magic around `__device__` to make the C++ compiler happy. But it still has holes like we found here.

Reply:
From my knowledge, depending on compilers, some of them compile all the code in the TU while some others check the syntax for all code in the TU but only compile the code used for object instantiation.
So maybe `__device__` code is present but is not used to instantiate anything so it could pass the compilation process.

Reply:
You are correct that the dead code optimizer (or the linker) could remove an unused method. However, we have definitely made debug builds of libcudf work at various points, and with `-O0` the dead code shouldn't be optimized out, so I don't think that can be the answer. I think @davidwendt's answer is correct though, that would just require defining the necessary macro as the empty string in that header file.